### ISSUE TITLE
Quarry timing out after 8-10  hours

### ISSUE THREAD

**Issue by Starmixcraft** [TheHolyWaffle/TeamSpeak-3-Java-API#285]
I have observd that the quarry sortof times out after a long time, no event is fired and no command can be sent, but the how am i command still runs with the message OK.

Log entry:

    [15:52:40] [DEBUG] [whoami] > whoami
    [15:52:40] [DEBUG] [whoami] < virtualserver_status=online virtualserver_id=1 virtualserver_unique_identifier=DfSfdKSgo3ftj+NDIHoSxb9KVuY= virtualserver_port=9987 client_id=7 client_channel_id=1 client_nickname=CloudSystem client_database_id=1 client_login_name=serveradmin client_unique_identifier=serveradmin client_origin_server_id=0
    [15:52:40] [DEBUG] [whoami] < error id=0 msg=ok

There is no error throwen wenn this happens

**Comment by rogermb** [comment]
Huh, that's strange.

If you take a look at the [KeepAliveThread](https://github.com/TheHolyWaffle/TeamSpeak-3-Java-API/blob/master/src/main/java/com/github/theholywaffle/teamspeak3/KeepAliveThread.java) source code, you'll see that it also just uses the API to send commands. In other words, it looks like you should be able to send commands, too.

Are you sure that you're not crashing your own program somehow without noticing?

But then again, that wouldn't explain why events aren't fired. Could you maybe upload your event listener code here so I could take a look at it?

**Comment by Starmixcraft** [comment]
Yes my programm runs 2 netty servers and thay function for weeks

**Comment by rogermb** [comment]
Alright, but that gives me pretty much nothing to work with. It sounds like it could be a deadlock or some other multi-threading issue somewhere in our code, but if the keep-alive still works, there's not really any place where such a bug could be hiding.

Essentially, if our keep-alive can use the API object to successfully send commands over and over again, there's no reason why you shouldn't be able to do the same.

Moreover, the only reason that the keep-alive thread is sending these commands is because no other commands have been sent for over 60 seconds. If we also consider that the command-sending thread can still properly handle the `whoami` command, the only conclusion we can make is that the command send queue isn't being filled because you're not making any calls to `TS3Api` or `TS3ApiAsync`.

I can't say why these API calls aren't being made, though. Maybe it's some kind of issue in `CommandFuture` that keeps you waiting for a TS3 command forever. But then again, that wouldn't explain why events aren't being fired anymore. Or could it be that you're only ever sending TS3 commands from inside event handlers?

If you're having this issue again, could you also fire up Java Mission Control (jmc.exe or just jmc), attach yourself to the no-longer-working program and create a thread dump by running the "Thread.print" diagnostic command? That way, if there's a deadlock, we'll be able to find it.

**Comment by Starmixcraft** [comment]
There is no deadlock, my program checks for that a runtime, i have done some more testing and it seams that the command system works but i wont send any commands because no event are fired, so the bug must be in the event notification, i also dont see the debug messages in the console if i switch the channel for example, i will provide a thread bump as soon as possible

**Comment by Starmixcraft** [comment]
https://hastebin.com/ovilesetaj.bash is a thread dump at the time the bug accured

**Comment by rogermb** [comment]
That... kinda explains everything. If you get kicked from the virtual server (e.g. because the virtual server is shut down / restarted), you'd also no longer receive any event notifications from that virtual server.

I've also looked through the thread dump you've sent me, and from what I can tell, everything seems to be working, too. The only threads that are blocked in our API are 1) the sender thread waiting for commands to send, 2) the receiver thread waiting for responses from the server, and 3) the keep-alive thread waiting for the timer to expire. In other words: everything is as it should be 😃 

The real question now is why you're getting kicked from that virtual server in the first place. It might be a good idea to monitor and log any `ClientLeaveEvent`s that mention your own client ID.

Alternatively, you could also periodically call `TS3Api#whoAmI` to check what virtual server you're currently on.

Strangely, however, the log entry of the `whoami` command you included in your first message says that you're still on the virtual server with ID 1 / port 9987. But if that were the case, you obviously wouldn't be getting the `invalid serverID` error message when trying to send a command. Did you copy that `whoami` command log from an earlier point in time when the query was still working?

**Comment by Starmixcraft** [comment]
@rogermb i just chcked the logs and the whoami command still works after the error, i dont understand wy?

**Comment by rogermb** [comment]
The whoami command works even if you're not on a virtual server, but the output should be different than what you've included in your first message. In particular, it should mention that you're now on the virtual server with the ID 0, which isn't actually a valid virtual server.

**Comment by Starmixcraft** [comment]
@rogermb, okay just check logs again, and you were correct,
`[01:23:56] [DEBUG] [whoami] > whoami
[01:23:56] [DEBUG] [whoami] < virtualserver_status=unknown virtualserver_unique_identifier virtualserver_port=0 virtualserver_id=0 client_id=0 client_channel_id=0 client_nickname client_database_id=0 client_login_name client_unique_identifier client_origin_server_id=0
[01:23:56] [DEBUG] [whoami] < error id=0 msg=ok` 
Now the question is how to prevent this

**Comment by rogermb** [comment]
(We could've saved so much time hunting down nonexistent multi-threading issues if we'd had the correct `whoami` command response from the get-go... 😛)

I think first of all, we need to find out why you're getting kicked from the virtual server in the first place.
If you check your old command log again, it would be interesting to know if there are any `notifyclientleftview` events right before the query stops working. If yes, could you send me the reason IDs and messages?

**Comment by Starmixcraft** [comment]
@rogermb 
The entire log is https://hastebin.com/tiqujaxawi.makefile i tryed to remove most of the nois but its still pritty long, maby you spot i dident

**Comment by Starmixcraft** [comment]
@rogermb 
Maby this is the problem, in my error logging system there is this error:

    An Error accured at 01:00:14 on the server CloudServer, a Error walkthrue folowes:
    java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:171)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.BufferedReader.fill(BufferedReader.java:161)
	at java.io.BufferedReader.readLine(BufferedReader.java:324)
	at java.io.BufferedReader.readLine(BufferedReader.java:389)
	at com.github.theholywaffle.teamspeak3.SocketReader.run(SocketReader.java:82)

**Comment by rogermb** [comment]
So in [this message](https://github.com/TheHolyWaffle/TeamSpeak-3-Java-API/issues/285#issuecomment-415775754) your `TS3Config` doesn't seem to have reconnecting turned on. But in your log, the query clearly attempts to reconnect. What's up with that?

And after you've reconnected, you don't re-run the required commands (logging in, selecting the virtual server, registering events, setting your nickname). That's why you're not on the correct virtual server.

If you have reconnecting turned on, you'll need to move your "startup" code (the setup commands I just mentioned) into the `onConnect` method of your `ConnectionHandler`. Please check out [this example](https://github.com/TheHolyWaffle/TeamSpeak-3-Java-API/blob/master/example/ReconnectExample.java).

**Comment by Starmixcraft** [comment]
I have now set up a custom reconnect handler that dose select a server and that stuff, but now i get this error: https://hastebin.com/bahosexibi.makefile

**Comment by rogermb** [comment]
Could I please see your new code where you set up your TS3Query, including the reconnect handler?
Please remove the username & password in the login command, but don't change anything else 😛

**Comment by Starmixcraft** [comment]
This is the new code i am usin: https://hastebin.com/ujipeduyat.java

**Comment by rogermb** [comment]
The exception will disappear if you use the synchronous API instead of the async API in the `onConnect` method. This will not block your application, as that handler is run in a thread from our thread pool.

I'm also not sure why you didn't follow the [example](https://github.com/TheHolyWaffle/TeamSpeak-3-Java-API/blob/master/example/ReconnectExample.java) I sent you earlier. You *can* create your own `ReconnectingConnectionHandler` and set all the parameters manually, but that's ugly as sin. Why not use something like `ts3Config.setReconnectStrategy(ReconnectStrategy.exponentialBackoff())` instead?

By the way, what you recreated by instantiating your own `ReconnectingConnectionHandler` is a reconnect strategy that first waits 50 milliseconds before attempting to reconnect. After that, it just tries to reconnect in an infinite loop without ever waiting, because you set the multiplier to 0. These sorts of mistakes is why `ReconnectStrategy` exists. I think what you were looking for was `ReconnectStrategy.constantBackoff(50)`.

But even that is too little delay in my opinion. Do you really think that if there's a connection problem, you should try to reconnect 20 times per second?

**Comment by Starmixcraft** [comment]
@rogermb if i use ReconnectStrategy.constantBackoff(50) it will do everything for me? like connect back to the server and register all events?

**Comment by rogermb** [comment]
No, the reconnect strategy isn't clairvoyant 😛 

Just use something like

```java
// One attempt to reconnect every second is more than enough
config.setReconnectStrategy(ReconnectStrategy.constantBackoff(1000));

config.setConnectionHandler(new ConnectionHandler() {

	@Override
	public void onConnect(TS3Query ts3Query) {
		TS3Api ts3Api = ts3Query.getApi();
		ts3Api.login(cloudConfig.getTs3user(), cloudConfig.getTs3pass());
		ts3Api.selectVirtualServerByPort(9987);
		ts3Api.registerAllEvents();
		ts3Api.addTS3Listeners(new TS3ListenerToEventManager());
		ts3Api.setNickname("CloudSystem");
	}

	@Override
	public void onDisconnect(TS3Query ts3Query) {
		// Nothing
	}
});
```

**Comment by rogermb** [comment]
Oh, actually, don't add your TS3 listener in the `onConnect` method. You only ever need to do that once, else you'll get duplicate events after reconnects. So just do that after you've called `ts3Query.connect()`.

**Comment by Starmixcraft** [comment]
Ok, thx for the good support, i probably never had figured that out on my own

**Comment by rogermb** [comment]
Should be fixed in [v1.2.0](https://github.com/TheHolyWaffle/TeamSpeak-3-Java-API/releases/tag/v1.2.0).

Thank you very much for reporting this issue! 😄
